### ISSUE TITLE
Facebook login not getting authenticated in Federated Identity

### ISSUE THREAD

**Issue by sourabhzynq** [aws-amplify/aws-sdk-ios#791]
### To help us solve your problem better, please answer the following list of questions.

* What service are you using?
AWS Cognito
* In what version of SDK are you facing the problem?
2.4.12
* Is the issue limited to Simulators / Actual Devices?
All
* Can your problem be resolved if you bump to a higher version of SDK?
Doesn't seem so
* Is this problem related to specific iOS version?
No
* How are you consuming the SDK? CocoaPods / Carthage / Prebuilt frameworks?
CocoaPods
* Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.

*Issue Description:

Hi Everyone,

I am integrating AWS with my iOS app for user sign in and sync. I have successfully implemented user sign up/sign in using user pool. Now I am integrating social providers sign ins with Federated IDs and User Pool. My code is in Objective C.

I have integrated FB and Google sign in with my app. It works well and I get tokens (FBAccessToken from FB). But this user is not getting authenticated in Federated IDs (It gives error message: 'Unauthenticated access is not supported for this identity pool.'). If I allow unauthenticated identities in my Fedreated ID setting, then this FB signed in user is shown as Unauthenticated Id in Federated ID, and obviously no related user in userPool.

How I implemented:

    I have added Facebook in Authenticated Providers for Federated ID.
    I have also added Facebook in Identity Providers for User Pool and done attributes mapping (mapped Id and Email).
    I have created a social ID provider class which implements AWSIdentityProviderManager protocol. This class contains logins method as given:

    (AWSTask<NSDictionary<NSString *, NSString *> *> )logins {
    NSLog(@"Inside logins method!");
    FBSDKAccessToken fbToken = [FBSDKAccessToken currentAccessToken];
    if(fbToken){
    NSString *token = fbToken.tokenString;
    return [AWSTask taskWithResult: @{ AWSIdentityProviderFacebook : token }];
    } else {
    NSLog(@"FB Token is not found in AWSSocialIdentityProvider class");
    return [AWSTask taskWithError:[NSError errorWithDomain:@"facebook"
    code:-1
    userInfo:@{@"error":@"No current Facebook access token"}]];
    }
    }

    My Sign In View Controller does FB sign in. It creates fbCredentialsProvider by passing the socialIDProvider class as an identityProvider in it's construction, implemented as below:

     self.fbCredentialsProvider = [[AWSCognitoCredentialsProvider alloc] initWithRegionType:AWSRegionUSEast1 identityPoolId:@"us-east-1:XXXXXXXXXXXX" identityProviderManager:self.socialIDProvider];
     AWSServiceConfiguration *configuration = [[AWSServiceConfiguration alloc] initWithRegion:AWSRegionUSEast1 credentialsProvider:self.fbCredentialsProvider];
     [AWSServiceManager defaultServiceManager].defaultServiceConfiguration = configuration;

     //wipe cached credentials
     [self.fbCredentialsProvider clearKeychain];
     [self.fbCredentialsProvider clearCredentials];

     //For calling fb logins
     [self.fbCredentialsProvider credentials];

    The above code is called in didCompleteWithResult method of FB Login Button.
    That's it.

I know this question has been asked by many and there are many issues/answers on forums, and I have read (I suppose) almost all of them, but none could help in resolving the issue.

Please help here as I am stuck at this issue for over a month, read all the AWS documention on AWS-iOS SDK, number of articles, forums and yet could not resolve it. Now we cannot afford to loose more time and have to finish it sooner, so need help from AWS team/fellow developers.

If anyone can point error in the code (if any), and provide correct code will be highly appreciated.

Thank you in advance!

**Comment by sourabhzynq** [comment]
@behrooziAWS @rohandubal  @karthiksaligrama

Hello friends,

Can anyone of you or others from AWS team kindly have a look at this and help?

**Comment by kvasukib** [comment]
Hi @sourabhzynq, Thanks for reporting to us. We are looking into it.

**Comment by behrooziAWS** [comment]
Do you see "Inside logins method" being printed?  Are you making calls to AWS services beyond what you've shown here?  Normally you configure your defaultServiceConfiguration once in the AppDelegate, not in didCompleteResult from the result of a button click.  If the Facebook access token is available prior to your logins method getting called and the first call to an aws service, you should get an authenticated id.

**Comment by sourabhzynq** [comment]
No, "Inside logins method" is not getting printed meaning the logins method is not getting called, even when I have a valid FB Access token. Let me give you brief idea about how I am using Cognito. My app doesn't require user to login immediately on launching the app. I have given an option to Sign Up/Sign In which a user can select from menu after he started using the app. For custom sign up/sign in, I am using Cognito UserPool and have created a manager (singleton) class which creates a userpool object. This pool object is used for Sign Up/Sign In workflow using Cognito UserPool. This worked well. Then I decided to implement authentication with social identity providers and Federated identities; FB Log In and Google Sign In. For startup, I tried integrating FB with Fed ID and above error is what I am getting. I wrote configuration code for FB in didCompleteResult because I already had written one configuration in the manager class (refer. above) for userpool. I am not sure if this approach is correct, you can suggest me on this.

**Comment by sourabhzynq** [comment]
@behrooziAWS  @kvasukib hi, any suggestions here?

**Comment by behrooziAWS** [comment]
Ah ok, so if you want to support both userpools and facebook in this manner, you still should only have 1 implementation of the AWSIdentityProviderManager protocol that you supply to the  AWSCognitoCredentialsProvider and provide a handle to your user pools object.

In your logins method, depending on how your user is authenticated you should return the appropriate logins map:

```
- (AWSTask<NSDictionary<NSString *, NSString *> *> *)logins {
    if(/*logic to determine if user is unauthenticated*/) {
        return [AWSTask taskWithResult:nil];
    }else if (/*logic to determine if user is Facebook*/){
        return [AWSTask taskWithResult: @{ AWSIdentityProviderFacebook : [FBSDKAccessToken currentAccessToken] }];
    }else {
        /*else delegate to user pools to generate the logins map*/
        return [self.userPool logins];
    }
}
```

**Comment by sourabhzynq** [comment]
Hi @behrooziAWS Thanks for your reply - and sorry that I am coming back so late.

**Comment by sourabhzynq** [comment]
Hi @behrooziAWS  Thanks for your reply - and sorry that I am coming back so late.
I tried it and could authenticate with Facebook and Google as well. But now when I am trying to use this authenticated user (immediately after the sign in) to invoke a lambda, I am getting an error from the lambda invoker as 'Authentication delegate not set'. Also when I am trying to get Identity Id and store it, it is not getting stored - though I could see the user being already authenticated. Please refer the snippet below:

```
//This is credential provider initialised once on app launch
		self.socialProvider =[[AWSSocialIdentityProvider alloc] init];
                 self.socialCredentialsProvider = [[AWSCognitoCredentialsProvider alloc]
                                                  initWithRegionType:AWSRegionUSEast1
                                                  identityPoolId:@"us-east-1:xxxxxxxxxxxxxx"
                                                  identityProviderManager:self.socialProvider];
		// Self.socialProvider is that class that implements AWSIdentityProviderManager protocol
                
                AWSServiceConfiguration *fbConfiguration = [[AWSServiceConfiguration alloc]
                                                            initWithRegion:AWSRegionUSEast1 credentialsProvider:self.socialCredentialsProvider];
                AWSServiceManager.defaultServiceManager.defaultServiceConfiguration = fbConfiguration;
                
                                
                [self.socialCredentialsProvider clearKeychain];
                
                FBSDKAccessToken* fbToken = [FBSDKAccessToken currentAccessToken];
                NSString *token = fbToken.tokenString;
                NSLog(@"FBToken is %@",token);
                
                            
                NSDictionary *dict = @{@"graph.facebook.com" : token};
                self.socialProvider.tokens = dict;

                 [self.socialCredentialsProvider clearCredentials];
  
                [[[self.socialCredentialsProvider credentials] continueWithBlock:^id _Nullable(AWSTask<NSArray *> * _Nonnull task) {
                    
                    NSLog(@"Inside credentials");
                    
                    NSArray *AWSCredentials = task.result;
		   //This does return credentials
                    NSLog(@"AWSCredentials is: %@",AWSCredentials);
                    return nil;
                    
                }]continueWithSuccessBlock:^id _Nullable(AWSTask<NSString *> * _Nonnull task) {
                    [[self.socialCredentialsProvider getIdentityId] continueWithBlock:^id(AWSTask *task) {
                        if (task.error) {
                            NSLog(@"Error: %@", task.error);
                        }
                        else {
                            
                            // the task result should contain the identity id
                            NSString* cognitoId = task.result;
			  //This does not return - in fact, this is not even getting printed in the console - means it is not getting executed ?
                            NSLog(@"Cognito Id contains: %@“,cognitoId);
                            
                        }
                        return nil;
                    }];
                   return nil;
               }];
//After this I call a function that invokes the lambda, which gives me the above mentioned error.
```

The above code gets executed on Facebook login.

Can you have a look at it and comment/guide.

**Comment by amit2908** [comment]
I too have the exactly same issue. Please help.

**Comment by rohandubal** [comment]
Hello @sourabhzynq 

Apologies for the delayed response.

The error you are seeing is due to the requirement of setting `interactiveAuthDelegate`. The expectation of the SDK is that you always set the `interactiveAuthDelegate` so that whenever there is a requirement of getting user authentication again we invoke it.

Also, FYI, we recently released an updated `AWSMobileClient` version `2.7.x` where we have declarative APIs for signIn / getTokens which also handle the refresh logic automatically. You can find documentation about it here: https://aws-amplify.github.io/docs/ios/authentication

Please let us know if this helps.

Thanks,
Rohan
